eis_toolkit/raster_processing/reclassify_raster.py

```diff
@@ -0,0 +1,511 @@
+    number_of_classes: int,
+    bands: Optional[Sequence[int]],
+) -> Tuple[Sequence[np.ndarray], dict]:
+
+    array_of_bands = []
+    out_image = []
+    out_meta = raster.meta.copy()
+
+    if bands is not None:
+        for band in raster.read(bands):
+            array_of_bands.append(band)
+    else:
+        array_of_bands = raster.read()
+        bands = np.arange(1, len(array_of_bands) + 1, 1).tolist()
+
+    for i in range(len(bands)):
+        data_array = array_of_bands[i]
+        breaks = mc.JenksCaspall(data_array, number_of_classes)
+        data = np.digitize(data_array, np.sort(breaks.bins))
+
+        out_image.append(data)
+
+    return out_image, out_meta
+
+
+@beartype
+def raster_with_natural_breaks(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader, number_of_classes: int, bands: Optional[Sequence[int]]
+) -> Tuple[Sequence[np.ndarray], dict]:
+    """Classify raster with natural breaks (Jenks Caspall).
+
+    If bands are not given, all bands are used for classification.
+
+    Args:
+        raster: Raster to be classified.
+        number_of_classes: The number of classes.
+        bands: Selected bands from multiband raster. Indexing begins from one. Defaults to None.
+
+    Returns:
+        Raster classified with natural breaks (Jenks Caspall) and metadata.
+
+    Raises:
+        InvalidParameterValueException: Bands is not a list, bands is not a list of integers,
+        bands exceeds the number of bands in the raster.
+    """
+    if bands is not None:
+        if not isinstance(bands, list):
+            raise InvalidParameterValueException("Expected bands parameter to be a list")
+        elif not all(isinstance(band, int) for band in bands):
+            raise InvalidParameterValueException("Expected bands to be a list of integers")
+        elif len(bands) > raster.count:
+            raise InvalidParameterValueException("The number of bands given exceeds the number of raster's bands")
+
+    out_image, out_meta = _raster_with_natural_breaks(raster, number_of_classes, bands)
+
+    return out_image, out_meta
+
+
+def _raster_with_geometrical_intervals(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader, number_of_classes: int, bands: Optional[Sequence[int]]
+) -> Tuple[Sequence[np.ndarray], dict]:
+
+    array_of_bands = []
+    out_image = []
+    out_meta = raster.meta.copy()
+
+    if bands is not None:
+        for band in raster.read(bands):
+            array_of_bands.append(band)
+    else:
+        array_of_bands = raster.read()
+
+    for i in range(len(array_of_bands)):
+
+        data_array = array_of_bands[i]
+        # Masking missing values with np.nan
+        data_array[data_array == -1.0e32] = np.nan
+
+        median_value = np.nanmedian(data_array)
+        max_value = np.nanmax(data_array)
+        min_value = np.nanmin(data_array)
+
+        data_array = np.ma.masked_where(np.isnan(data_array), data_array)
```

Comment:
Is this line needed? I do not see the array mask being used, instead the array is compared to `np.nan` everywhere anyway instead of using `data_array.mask` to see what is `np.nan`.